Pip install failing on Ubuntu 12.04, Python 3.4.3

This is the first time I have attempted to utilize this library, so hopefully I am missing something easy.

I installed the dependencies:

`apt-get install libxml2-dev libxmlsec1-dev`

A `pip install xmlsec==0.3.3` generates the following error:

```
[...]
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security -D_FORTIFY_SOURCE=2 -fPIC -DXMLSEC_NO_GOST=1 -D__XMLSEC_FUNCTION__=__FUNCTION__ -DXMLSEC_OPENSSL_098=1 -DXMLSEC_CRYPTO_OPENSSL=1 -DXMLSEC_NO_XKMS=1 -DXMLSEC_CRYPTO=\\openssl\\ -DXMLSEC_NO_SIZE_T= -DXMLSEC_NO_CRYPTO_DYNAMIC_LOADING=1 -Isrc -I/home/daaray/.virtualenvs/market_builder/build/xmlsec/lxml-3.5.0b1-py3.4-linux-x86_64.egg/lxml/includes -I/home/daaray/.virtualenvs/market_builder/build/xmlsec/lxml-3.5.0b1-py3.4-linux-x86_64.egg/lxml -I/usr/include/xmlsec1 -I/usr/include/libxml2 -I/usr/include/python3.4m -I/home/daaray/.virtualenvs/market_builder/include/python3.4m -c src/xmlsec/constants.c -o build/temp.linux-x86_64-3.4/src/xmlsec/constants.o -DXMLSEC_CRYPTO_OPENSSL=1

src/xmlsec/constants.c: In function ‘PyInit_constants’:

src/xmlsec/constants.c:2419:61: error: ‘xmlSecTransformEcdsaSha1Id’ undeclared (first use in this function)

src/xmlsec/constants.c:2419:61: note: each undeclared identifier is reported only once for each function it appears in

src/xmlsec/constants.c:2431:61: error: ‘xmlSecTransformEcdsaSha224Id’ undeclared (first use in this function)

src/xmlsec/constants.c:2443:61: error: ‘xmlSecTransformEcdsaSha256Id’ undeclared (first use in this function)

src/xmlsec/constants.c:2455:61: error: ‘xmlSecTransformEcdsaSha384Id’ undeclared (first use in this function)

src/xmlsec/constants.c:2467:61: error: ‘xmlSecTransformEcdsaSha512Id’ undeclared (first use in this function)

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

Please advise.

Reporting back...with `gcc v4.6.3` a `pip install xmlsec==0.3.1` resulted in the following for me:

```
src/xmlsec/enc.c: In function '__pyx_import_star_set':

src/xmlsec/enc.c:3384:37: error: lvalue required as left operand of assignment

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

In this case, Its seems that the problem is related to lxml 3.5.0:
```
/tmp/easy_install-Ew8_x0/lxml-3.5.0/temp/xmlXPathInitI_sX_B.c:2:1: warning: return type defaults to ‘int’ [-Wimplicit-int]
 main (int argc, char **argv) {
```

--- a/src/xmlsec.h
+++ b/src/xmlsec.h
@@ -1,7 +1,9 @@
-#include <xmlsec/xmltree.h>
+#include <xmlsec/xmlsec.h>
+#include <xmlsec/errors.h>
+#include <xmlsec/templates.h>
 #include <xmlsec/xmldsig.h>
 #include <xmlsec/xmlenc.h>
-#include <xmlsec/templates.h>
-#include <xmlsec/errors.h>
+#include <xmlsec/xmltree.h>
 #include <xmlsec/openssl/app.h>
 #include <xmlsec/openssl/crypto.h>
+#include <xmlsec/openssl/x509.h>

--- a/src/xmlsec/key.pxd
+++ b/src/xmlsec/key.pxd
@@ -120,4 +120,3 @@ cdef class Key(object):
 
 cdef class KeysManager(object):
     cdef xmlSecKeysMngrPtr _handle
-    cdef bint _mode

--- a/src/xmlsec/key.pxd
+++ b/src/xmlsec/key.pxd
@@ -98,6 +98,8 @@ cdef extern from "xmlsec.h":  # xmlsec/keys.h
 
     int xmlSecOpenSSLAppDefaultKeysMngrInit(xmlSecKeysMngrPtr) nogil
 
+    int xmlSecOpenSSLKeysMngrInit(xmlSecKeysMngrPtr) nogil
+
     int xmlSecOpenSSLAppDefaultKeysMngrAdoptKey(xmlSecKeysMngrPtr, xmlSecKeyPtr) nogil
 
     int xmlSecOpenSSLAppKeysMngrCertLoad(
@@ -118,3 +120,4 @@ cdef class Key(object):
 
 cdef class KeysManager(object):
     cdef xmlSecKeysMngrPtr _handle
+    cdef bint _mode

--- a/src/xmlsec/key.pyx
+++ b/src/xmlsec/key.pyx
@@ -235,7 +235,6 @@ cdef class KeysManager(object):
         cdef int rv
         cdef xmlSecKeysMngrPtr handle
 
-        self._mode = 0
         handle = xmlSecKeysMngrCreate()
         if handle == NULL:
             raise InternalError("failed to create keys manager", -1)

--- a/src/xmlsec/key.pyx
+++ b/src/xmlsec/key.pyx
@@ -235,18 +235,32 @@ cdef class KeysManager(object):
         cdef int rv
         cdef xmlSecKeysMngrPtr handle
 
+        self._mode = 0
         handle = xmlSecKeysMngrCreate()
         if handle == NULL:
             raise InternalError("failed to create keys manager", -1)
-
-        rv = xmlSecOpenSSLAppDefaultKeysMngrInit(handle)
-        if rv < 0:
-            raise InternalError("failed to initialize keys manager", rv)
         self._handle = handle
 
     def __dealloc__(self):
         if self._handle != NULL:
             xmlSecKeysMngrDestroy(self._handle)
+
+
+    def __ensure_init(self, expected_mode):
+        if self._mode == expected_mode:
+            return
+
+        if self._mode != 0:
+           raise InternalError("Cannot use one manager for verify and sign", -1)
+
+        if expected_mode == 1:
+           rv = xmlSecOpenSSLAppDefaultKeysMngrInit(self._handle)
+        else:
+           rv = xmlSecOpenSSLKeysMngrInit(self._handle)
+
+        if rv < 0:
+            raise InternalError("failed to initialize keys manager", rv)
+
 
     def add_key(self, Key key):
         """add (a copy of) *key*."""
@@ -259,6 +273,7 @@ cdef class KeysManager(object):
         if key_handle == NULL:
             raise InternalError("failed to copy key", -1)
 
+        self.__ensure_init(1)
         rv = xmlSecOpenSSLAppDefaultKeysMngrAdoptKey(self._handle, key_handle)
         if rv < 0:
             xmlSecKeyDestroy(key_handle)
@@ -272,6 +287,7 @@ cdef class KeysManager(object):
         cdef int rv
         cdef const_char* c_filename = <const_char*>_b(filename)
 
+        self.__ensure_init(2)
         with nogil:
             rv = xmlSecOpenSSLAppKeysMngrCertLoad(self._handle, c_filename, format, type)
 
@@ -292,6 +308,7 @@ cdef class KeysManager(object):
         c_size = len(data)
         c_data = <const_unsigned_char*><char*>data
 
+        self.__ensure_init(2)
         with nogil:
             rv = xmlSecOpenSSLAppKeysMngrCertLoadMemory(self._handle, c_data, c_size, format, type)
 

--- a/src/xmlsec/key.pyx
+++ b/src/xmlsec/key.pyx
@@ -239,28 +239,16 @@ cdef class KeysManager(object):
         handle = xmlSecKeysMngrCreate()
         if handle == NULL:
             raise InternalError("failed to create keys manager", -1)
+        rv = xmlSecOpenSSLAppDefaultKeysMngrInit(handle)
+        if rv < 0:
+            xmlSecKeysMngrDestroy(handle)
+            raise InternalError("failed to init manager", rv)
+
         self._handle = handle
 
     def __dealloc__(self):
         if self._handle != NULL:
             xmlSecKeysMngrDestroy(self._handle)
-
-
-    def __ensure_init(self, expected_mode):
-        if self._mode == expected_mode:
-            return
-
-        if self._mode != 0:
-           raise InternalError("Cannot use one manager for verify and sign", -1)
-
-        if expected_mode == 1:
-           rv = xmlSecOpenSSLAppDefaultKeysMngrInit(self._handle)
-        else:
-           rv = xmlSecOpenSSLKeysMngrInit(self._handle)
-
-        if rv < 0:
-            raise InternalError("failed to initialize keys manager", rv)
-
 
     def add_key(self, Key key):
         """add (a copy of) *key*."""
@@ -273,7 +261,6 @@ cdef class KeysManager(object):
         if key_handle == NULL:
             raise InternalError("failed to copy key", -1)
 
-        self.__ensure_init(1)
         rv = xmlSecOpenSSLAppDefaultKeysMngrAdoptKey(self._handle, key_handle)
         if rv < 0:
             xmlSecKeyDestroy(key_handle)
@@ -287,7 +274,6 @@ cdef class KeysManager(object):
         cdef int rv
         cdef const_char* c_filename = <const_char*>_b(filename)
 
-        self.__ensure_init(2)
         with nogil:
             rv = xmlSecOpenSSLAppKeysMngrCertLoad(self._handle, c_filename, format, type)
 
@@ -308,7 +294,6 @@ cdef class KeysManager(object):
         c_size = len(data)
         c_data = <const_unsigned_char*><char*>data
 
-        self.__ensure_init(2)
         with nogil:
             rv = xmlSecOpenSSLAppKeysMngrCertLoadMemory(self._handle, c_data, c_size, format, type)
 

--- a/src/xmlsec/utils.pyx
+++ b/src/xmlsec/utils.pyx
@@ -50,4 +50,4 @@ def shutdown():
 
 
 def enable_debug_trace(flag):
-    xmlSecErrorsDefaultCallbackEnableOutput(<int>flag)
+    xmlSecErrorsDefaultCallbackEnableOutput(1 if flag else 0)

--- a/src/xmlsec/utils.pyx
+++ b/src/xmlsec/utils.pyx
@@ -16,16 +16,16 @@ def init():
     This is called upon library import and does not need to be called
     again (unless @ref _shutdown is called explicitly).
     """
-    r = xmlSecInit()
-    if r != 0:
+    if xmlSecInit() < 0:
         return False
 
-    r = xmlSecOpenSSLInit()
-    if r != 0:
+    if xmlSecOpenSSLAppInit(NULL) < 0:
+        xmlSecShutdown()
         return False
 
-    r = xmlSecOpenSSLAppInit(NULL)
-    if r != 0:
+    if xmlSecOpenSSLInit() < 0:
+        xmlSecOpenSSLAppShutdown()
+        xmlSecShutdown()
         return False
 
     return True
@@ -37,16 +37,16 @@ def shutdown():
     This is called automatically upon interpreter termination and
     should not need to be called explicitly.
     """
-    r = xmlSecOpenSSLAppShutdown()
-    if r != 0:
+    if xmlSecOpenSSLShutdown() < 0:
         return False
 
-    r = xmlSecOpenSSLShutdown()
-    if r != 0:
+    if xmlSecOpenSSLAppShutdown() < 0:
         return False
 
-    r = xmlSecShutdown()
-    return r == 0
+    if xmlSecShutdown() < 0:
+        return False
+
+    return True
 
 
 def enable_debug_trace(flag):

--- a/tests/examples/base.py
+++ b/tests/examples/base.py
@@ -18,4 +18,3 @@ def compare(name, result):
 
     # Compare the results.
     assert expected_text == result_text
-

--- a/tests/examples/sign4-doc.xml
+++ b/tests/examples/sign4-doc.xml
@@ -1,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+XML Security Library example: Original XML doc file for sign3 example. 
+-->
+<Envelope xmlns="urn:envelope" ID="ef115a20-cf73-11e5-aed1-3c15c2c2cc88">
+  <Data>
+	Hello, World!
+  </Data>
+</Envelope>

--- a/tests/examples/sign4-res.xml
+++ b/tests/examples/sign4-res.xml
@@ -1,0 +1,55 @@
+<Envelope xmlns="urn:envelope" ID="ef115a20-cf73-11e5-aed1-3c15c2c2cc88">
+  <Data>
+	Hello, World!
+  </Data>
+<ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
+<ds:SignedInfo>
+<ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
+<ds:SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
+<ds:Reference URI="#ef115a20-cf73-11e5-aed1-3c15c2c2cc88">
+<ds:Transforms>
+<ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
+<ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
+</ds:Transforms>
+<ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
+<ds:DigestValue>ERS7F/ifxZoyTe0mhco+HeAEKGo=</ds:DigestValue>
+</ds:Reference>
+</ds:SignedInfo>
+<ds:SignatureValue>G+mZNFqh9w0wIkDSbuYwvVDu7CMP8PEsw7jfwiZBC8nyF3loAtYKKAkdi6Zy3dJs
+tU8qKfhzvabmCjdIrGkFTdtlCNCVKDMzwogFtxEX4Oh77X6jjx4b22XNJx4AbnUG
+JV/EcsD+po8s5qVEXw62lRRd8cMDafbzOA/rBH96CMNgZhzxyaF9VRLa/vbt1ht2
+hE1KkdZCB4Y0Lv3QyeDL2jax3NFks9FUv8IqoWYQSvywdMLY2ZMiQ9UpPeVfMizi
+trd5zDUSD/s3hyIEs4gD5NJF3HZPD/Fe2Zw1PBPj9eLADdEzcdueyCdPJ2YioFIi
+1sMW/qPDhR/DoOJwGpUxwQ==</ds:SignatureValue>
+<ds:KeyInfo>
+<ds:X509Data>
+<ds:X509Certificate>MIIE3zCCBEigAwIBAgIBBTANBgkqhkiG9w0BAQQFADCByzELMAkGA1UEBhMCVVMx
+EzARBgNVBAgTCkNhbGlmb3JuaWExEjAQBgNVBAcTCVN1bm55dmFsZTE9MDsGA1UE
+ChM0WE1MIFNlY3VyaXR5IExpYnJhcnkgKGh0dHA6Ly93d3cuYWxla3NleS5jb20v
+eG1sc2VjKTEZMBcGA1UECxMQUm9vdCBDZXJ0aWZpY2F0ZTEWMBQGA1UEAxMNQWxl
+a3NleSBTYW5pbjEhMB8GCSqGSIb3DQEJARYSeG1sc2VjQGFsZWtzZXkuY29tMB4X
+DTAzMDMzMTA0MDIyMloXDTEzMDMyODA0MDIyMlowgb8xCzAJBgNVBAYTAlVTMRMw
+EQYDVQQIEwpDYWxpZm9ybmlhMT0wOwYDVQQKEzRYTUwgU2VjdXJpdHkgTGlicmFy
+eSAoaHR0cDovL3d3dy5hbGVrc2V5LmNvbS94bWxzZWMpMSEwHwYDVQQLExhFeGFt
+cGxlcyBSU0EgQ2VydGlmaWNhdGUxFjAUBgNVBAMTDUFsZWtzZXkgU2FuaW4xITAf
+BgkqhkiG9w0BCQEWEnhtbHNlY0BhbGVrc2V5LmNvbTCCASIwDQYJKoZIhvcNAQEB
+BQADggEPADCCAQoCggEBAJe4/rQ/gzV4FokE7CthjL/EXwCBSkXm2c3p4jyXO0Wt
+quaNC3dxBwFPfPl94hmq3ZFZ9PHPPbp4RpYRnLZbRjlzVSOq954AXOXpSew7nD+E
+mTqQrd9+ZIbGJnLOMQh5fhMVuOW/1lYCjWAhTCcYZPv7VXD2M70vVXDVXn6ZrqTg
+qkVHE6gw1aCKncwg7OSOUclUxX8+Zi10v6N6+PPslFc5tKwAdWJhVLTQ4FKG+F53
+7FBDnNK6p4xiWryy/vPMYn4jYGvHUUk3eH4lFTCr+rSuJY8i/KNIf/IKim7g/o3w
+Ae3GM8xrof2mgO8GjK/2QDqOQhQgYRIf4/wFsQXVZcMCAwEAAaOCAVcwggFTMAkG
+A1UdEwQCMAAwLAYJYIZIAYb4QgENBB8WHU9wZW5TU0wgR2VuZXJhdGVkIENlcnRp
+ZmljYXRlMB0GA1UdDgQWBBQkhCzy1FkgYosuXIaQo6owuicanDCB+AYDVR0jBIHw
+MIHtgBS0ue+a5pcOaGUemM76VQ2JBttMfKGB0aSBzjCByzELMAkGA1UEBhMCVVMx
+EzARBgNVBAgTCkNhbGlmb3JuaWExEjAQBgNVBAcTCVN1bm55dmFsZTE9MDsGA1UE
+ChM0WE1MIFNlY3VyaXR5IExpYnJhcnkgKGh0dHA6Ly93d3cuYWxla3NleS5jb20v
+eG1sc2VjKTEZMBcGA1UECxMQUm9vdCBDZXJ0aWZpY2F0ZTEWMBQGA1UEAxMNQWxl
+a3NleSBTYW5pbjEhMB8GCSqGSIb3DQEJARYSeG1sc2VjQGFsZWtzZXkuY29tggEA
+MA0GCSqGSIb3DQEBBAUAA4GBALU/mzIxSv8vhDuomxFcplzwdlLZbvSQrfoNkMGY
+1UoS3YJrN+jZLWKSyWE3mIaPpElqXiXQGGkwD5iPQ1iJMbI7BeLvx6ZxX/f+c8Wn
+ss0uc1NxfahMaBoyG15IL4+beqO182fosaKJTrJNG3mc//ANGU9OsQM9mfBEt4oL
+NJ2D</ds:X509Certificate>
+</ds:X509Data>
+</ds:KeyInfo>
+</ds:Signature></Envelope>

--- a/tests/examples/test_sign.py
+++ b/tests/examples/test_sign.py
@@ -158,6 +158,70 @@ def test_sign_generated_template_pem_with_x509():
     compare('sign3-res.xml', template)
 
 
+def test_sign_generated_template_pem_with_x509_with_custom_ns():
+    """
+    Should sign a file using a dynamicaly created template, key from PEM
+    file and an X509 certificate with custom ns.
+    """
+
+    # Load document file.
+    template = parse_xml('sign4-doc.xml')
+    xmlsec.tree.add_ids(template, ["ID"])
+    elem_id = template.get('ID', None)
+    if elem_id:
+        elem_id = '#' + elem_id
+    # Create a signature template for RSA-SHA1 enveloped signature.
+    signature_node = xmlsec.template.create(
+        template,
+        xmlsec.Transform.EXCL_C14N,
+        xmlsec.Transform.RSA_SHA1, ns='ds')
+
+    assert signature_node is not None
+
+    # Add the <ds:Signature/> node to the document.
+    template.append(signature_node)
+
+    # Add the <ds:Reference/> node to the signature template.
+    ref = xmlsec.template.add_reference(signature_node, xmlsec.Transform.SHA1, uri=elem_id)
+
+    # Add the enveloped transform descriptor.
+    xmlsec.template.add_transform(ref, xmlsec.Transform.ENVELOPED)
+    # Add the excl_c14n transform descriptor.
+    xmlsec.template.add_transform(ref, xmlsec.Transform.EXCL_C14N)
+
+    # Add the <ds:KeyInfo/> and <ds:KeyName/> nodes.
+    key_info = xmlsec.template.ensure_key_info(signature_node)
+    xmlsec.template.add_x509_data(key_info)
+
+    # Create a digital signature context (no key manager is needed).
+    ctx = xmlsec.SignatureContext()
+
+    # Load private key (assuming that there is no password).
+    filename = path.join(BASE_DIR, 'rsakey.pem')
+    key = xmlsec.Key.from_file(filename, xmlsec.KeyFormat.PEM)
+
+    assert key is not None
+
+    # Load the certificate and add it to the key.
+    filename = path.join(BASE_DIR, 'rsacert.pem')
+    key.load_cert_from_file(filename, xmlsec.KeyFormat.PEM)
+
+    # Set key name to the file name (note: this is just a test).
+    key.name = path.basename(filename)
+
+    # Set the key on the context.
+    ctx.key = key
+
+    assert ctx.key is not None
+    assert ctx.key.name == path.basename(filename)
+
+    # Sign the template.
+    ctx.sign(signature_node)
+
+    # Assert the contents of the XML document against the expected result.
+    compare('sign4-res.xml', template)
+
+
 def test_sign_binary():
     ctx = xmlsec.SignatureContext()
     filename = path.join(BASE_DIR, 'rsakey.pem')

--- a/tests/examples/test_verify.py
+++ b/tests/examples/test_verify.py
@@ -4,14 +4,14 @@ import xmlsec
 from .base import parse_xml, BASE_DIR
 
 
-@mark.parametrize('index', range(1, 4))
+@mark.parametrize('index', range(1, 5))
 def test_verify_with_pem_file(index):
     """Should verify a signed file using a key from a PEM file.
     """
 
     # Load the XML document.
     template = parse_xml('sign%d-res.xml' % index)
-
+    xmlsec.tree.add_ids(template, ["ID"])
     # Find the <Signature/> node.
     signature_node = xmlsec.tree.find_node(template, xmlsec.Node.SIGNATURE)
 

